File: Module_1/data_race.cpp
#include <thread>
#include <iostream>

int counter = 0; // Shared mutable state — danger!

void increment() {
    for (int i = 0; i < 1000000; i++)
        counter++; // RACE CONDITION: read-modify-write is not atomic
}

int main() {
    std::thread t1(increment);
    std::thread t2(increment);
    t1.join(); t2.join();
    // Expected: 2,000,000  Actual: ???
    std::cout << counter << std::endl;
}

File: Module_1/data_race.rs
use std::thread;

fn main() {
    let mut counter = 0;

    let t1 = thread::spawn(move || {
        for _ in 0..1_000_000 {
            counter += 1; // This is a copy, not a reference)
        }
    });

    let t2 = thread::spawn(move || {
        for _ in 0..1_000_000 {
            counter += 1; // This is a copy, not a reference)
        }
    });

    t1.join().unwrap();
    t2.join().unwrap();
    println!("Result: {}", counter); 
}